backend/compact-connect/lambdas/python/provider-data-v1/handlers/privileges.py

```
@@ -64,4 +64,50 @@ def deactivate_privilege(event: dict, context: LambdaContext):  # noqa: ARG001 u
+            else:
+                # Send notification to the provider
+                logger.info(
+                    'Sending privilege deactivation notification to provider',
+                    provider_id=provider_id,
+                    provider_email=provider_email,
+                )
+                config.email_service_client.send_provider_privilege_deactivation_email(
+                    compact=compact,
+                    provider_email=provider_email,
+                    privilege_id=deactivated_privilege_id,
+                )
+        except CCInternalException as e:
+            # Log the error but don't fail the deactivation process
+            logger.error('Failed to send provider privilege deactivation notifications', exception=str(e))
+            failed_to_send_notification = True
+
+        try:
+            # Send notification to the jurisdiction
+            logger.info('Sending privilege deactivation notification to jurisdiction', jurisdiction=jurisdiction)
+            provider_first_name = provider['givenName']
+            provider_last_name = provider['familyName']
+            config.email_service_client.send_jurisdiction_privilege_deactivation_email(
+                compact=compact,
+                jurisdiction=jurisdiction,
+                privilege_id=deactivated_privilege_id,
+                provider_first_name=provider_first_name,
+                provider_last_name=provider_last_name,
+            )
+        except CCInternalException as e:
+            # Log the error but don't fail the deactivation process
+            logger.error('Failed to send jurisdiction privilege deactivation notifications', exception=str(e))
+            failed_to_send_notification = True
+
+    if failed_to_send_notification:
+        raise CCInternalException('Privilege was deactivated, but email notification failed to send.')
```

Do we really want to throw an error if this happens? Will the frontend see this message? If not, what indication will they have that the failed sending of the email is the only issue?

---

This is a good point. The main purpose of raising the exception here was to ensure we have an alert that fires as this requires investigation into the cause of the failure. Given that this will result in a 502, unfortunately the user will not see this particular message, but if they try again, they will get back a 400 with the message 'Privilege already deactivated', and if they refresh the page they will see that the privilege is in an inactive status.

The alternative is to return a 200, which doesn't seem right either, since the full handler was not successful. If we did return a 200 though, we could send a metric with an associated cloudwatch alarm that will fire if this occurs so support can still investigate. Curious what your thoughts are for how we should approach error handling here.

---

Got it. Ultimately, I defer to you and think either choice is reasonable. My only thought is regarding the 'importance' of the email to the business logic.

If the emails _were_ vital, I would expect the handler to raise an exception if it failed, _and_ perhaps to rollback all other logic (i.e., reactivate the privilege).

If it were less vital, I might expect the handler to follow your 2nd suggested pattern – return 200, log error separately as a custom metric.

It seems overly complicated to implement rollback logic, in which case I would opt for the 200 approach. But that's assuming the above is good logic, which may or may not be the case.

---

We can move to the 200 approach, since in either case if the email notification step fails the end user can't do anything about it. Support staff will need to figure out what went wrong and likely resend the email notification once the issue has been addressed. We expect this occurrence to be extremely rare, so I will set up a custom metric with a cloudwatch alert, and just return a 200

---

Update has been pushed

---

Verified in my sandbox env that the alert fires when this exception case is triggered